http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/ResponseHandler.java

```diff
@@ -382,28 +435,59 @@ public void cancel() {
         }
     }
 
-    private Throwable wrapException(Throwable originalCause) {
-        if (originalCause instanceof ReadTimeoutException) {
-            return new IOException("Read timed out", originalCause);
-        } else if (originalCause instanceof WriteTimeoutException) {
-            return new IOException("Write timed out", originalCause);
-        }
-
-        return originalCause;
-    }
-
     private void notifyIfResponseNotCompleted(ChannelHandlerContext handlerCtx) {
         RequestContext requestCtx = handlerCtx.channel().attr(REQUEST_CONTEXT_KEY).get();
         Boolean responseCompleted = handlerCtx.channel().attr(RESPONSE_COMPLETE_KEY).get();
         Boolean lastHttpContentReceived = handlerCtx.channel().attr(LAST_HTTP_CONTENT_RECEIVED_KEY).get();
         handlerCtx.channel().attr(KEEP_ALIVE).set(false);
 
         if (!Boolean.TRUE.equals(responseCompleted) && !Boolean.TRUE.equals(lastHttpContentReceived)) {
-            IOException err = new IOException("Server failed to send complete response. " + CLOSED_CHANNEL_MESSAGE);
+            IOException err = new IOException(NettyUtils.closedChannelMessage(handlerCtx.channel()));
             runAndLogError(handlerCtx.channel(), "Fail to execute SdkAsyncHttpResponseHandler#onError",
                            () -> requestCtx.handler().onError(err));
             executeFuture(handlerCtx).completeExceptionally(err);
             runAndLogError(handlerCtx.channel(), "Could not release channel", () -> closeAndRelease(handlerCtx));
         }
     }
+
+    private static final class DataCountingPublisher implements Publisher<ByteBuffer> {
+        private final ChannelHandlerContext ctx;
+        private final Publisher<ByteBuffer> delegate;
```

Comment:
Nit: Maybe we could consider trying to reuse [NotifyingPublisher](https://github.com/aws/aws-sdk-java-v2/blob/738f53b7e30f65080c5194d9e879e5235b5e85d5/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/progress/NotifyingAsyncResponseTransformer.java#L69-L118) here.

We could create a publisher-specific listener interface, and `AsyncResponseTransformerListener` would then extend that, so this code would only depend on the publisher-specific listener.

Reply:
Sorry, I'm not familiar with NotifyingPublisher. What would be the benefit to adding the additional layer of abstraction provide, since Publisher is a well known abstraction? Where would we put that shared code (since AsyncResponseTransformerListener and NotifyingPublisher aren't in netty's dependency closure)?

Reply:
It's been a reoccurring pattern to wrap a publisher/subscriber with a delegating implementation simply to listen in on events. Something like `NotifyingPublisher` would reduce boilerplate code for this common use case and make it less error prone. Just a nit/idea, we can defer this to future refactoring.

Reply:
SdkPublisher is probably the place to put such a construct, imo. Unfortunately netty wouldn't have access to that, since it's in sdk-core.